src/platforms/esp32/main/platform_nifs.c

```diff
@@ -133,6 +140,194 @@ static term nif_esp_reset_reason(Context *ctx, int argc, term argv[])
+    nvs_close(nvs);
+    switch (err) {
+        case ESP_OK:
+            TRACE("Wrote blob to NVS. namespace '%s' key '%s' size: %i\n", namespace, key, size);
+            return OK_ATOM;
+        default:
+            fprintf(stderr, "Unable to set NVS blob. namespace='%s' key='%s' size=%i err=%i\n", namespace, key, size, err);
+            RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+static term nif_esp_nvs_erase_key(Context *ctx, int argc, term argv[])
+{
+    VALIDATE_VALUE(argv[0], term_is_atom);
+    VALIDATE_VALUE(argv[1], term_is_atom);
+
+    char namespace[MAX_NVS_KEY_SIZE + 1];
+    if (write_atom_c_string(ctx, namespace, MAX_NVS_KEY_SIZE + 1, argv[0]) != 0) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    char key[MAX_NVS_KEY_SIZE + 1];
+    if (write_atom_c_string(ctx, key, MAX_NVS_KEY_SIZE + 1, argv[1]) != 0) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    nvs_handle nvs;
+    esp_err_t err = nvs_open(namespace, NVS_READWRITE, &nvs);
+    switch (err) {
+        case ESP_OK:
+            break;
+        default:
+            fprintf(stderr, "Unable to open NVS. namespace '%s' err=%i\n", namespace, err);
+            RAISE_ERROR(BADARG_ATOM);
+    }
+
+    err = nvs_erase_key(nvs, key);
+    nvs_close(nvs);
+    switch (err) {
+        case ESP_OK:
+            TRACE("Erased key. namespace '%s' key '%s'\n", namespace, key);
+            return OK_ATOM;
+        case ESP_ERR_NVS_NOT_FOUND:
+            TRACE("No such entry -- ok.  namespace='%s' key='%s'\n", namespace, key);
+            return OK_ATOM;
+        default:
+            fprintf(stderr, "Unable to erase key. namespace='%s' key='%s' err=%i\n", namespace, key, err);
+            RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+static term nif_esp_nvs_erase_all(Context *ctx, int argc, term argv[])
+{
+    VALIDATE_VALUE(argv[0], term_is_atom);
+
+    char namespace[MAX_NVS_KEY_SIZE + 1];
+    if (write_atom_c_string(ctx, namespace, MAX_NVS_KEY_SIZE + 1, argv[0]) != 0) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    nvs_handle nvs;
+    esp_err_t err = nvs_open(namespace, NVS_READWRITE, &nvs);
+    switch (err) {
+        case ESP_OK:
+            break;
+        default:
+            fprintf(stderr, "Unable to open NVS. namespace '%s' err=%i\n", namespace, err);
+            RAISE_ERROR(BADARG_ATOM);
+    }
+
+    err = nvs_erase_all(nvs);
+    nvs_close(nvs);
+    switch (err) {
+        case ESP_OK:
+            TRACE("Erased all. namespace '%s'\n", namespace);
+            return OK_ATOM;
+        default:
+            fprintf(stderr, "Unable to erase all. namespace='%s' err=%i\n", namespace, err);
+            RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+static term nif_esp_nvs_reformat(Context *ctx, int argc, term argv[])
+{
+    esp_err_t err = nvs_flash_erase();
+    switch (err) {
+        case ESP_OK:
+            break;
+        default:
+            fprintf(stderr, "Unable to reformat NVS partition. err=%i\n", err);
+            RAISE_ERROR(BADARG_ATOM);
```

Comment:
nev_reformat does not takes arguments, so it would be odd to return a badarg atom as an error. I would chose a different error.

Reply:
Good point(s).  There are a lot of potential errors that could be encoded as atoms (https://github.com/espressif/esp-idf/blob/ca1e5e5bc52182a577f1706c8dcbb4d9d7afe310/components/nvs_flash/include/nvs.h#L32), but I don't know how extensively we should go with this.  Perhaps returning the error code would be sufficient.  (I am nervous about adding more and more code and atoms to handle esoteric errors on a limited platform like ESP)